File: UseCases/tests/codes/seanhenry/mockgenerator/util/UniqueMethodNameGeneratorTests.java
package codes.seanhenry.mockgenerator.util;

import junit.framework.TestCase;

public class UniqueMethodNameGeneratorTests extends TestCase {

  private UniqueMethodNameGenerator generator;

  public void test_uniqueMethodName_shouldReturnMethodName() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("methodName")
    }, new String[]{
      "methodName"
    });
  }

  public void test_shouldReturnNull_whenIDDoesNotExist() throws Exception {
    generator = createGenerator();
    assertNull(generator.getMethodName("1"));
  }

  public void test_uniqueMethodName_shouldReturnMethodName_whenNoOverloadedMethods() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("methodName", "param", "param2")
    }, new String[]{
      "methodName" // should not add labels when method name is unique
    });
  }

  public void test_duplicateMethodName_shouldAppendFirstParameterLabel_toMethodName() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("methodName"),
      new MethodModel("anotherMethod"),
      new MethodModel("methodName", "param: Type")
    }, new String[]{
      "methodName",
      "anotherMethod",
      // should add label when method name is overloaded
      // should not use type when label is unique
      "methodNameParam"
    });
  }

  public void test_duplicateMethodName_shouldAppendFirstParameterLabel_toMethodName_whenGivenMethod_hasMultipleParamLabels() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("animate"),
      new MethodModel("animate", "withDuration duration: Type"),
      new MethodModel("animate", "withDuration duration: Type", "delay: Type"),
      new MethodModel("animate", "withDuration duration: Type", "delay: Type", "easing: Ease")
    }, new String[]{
      "animate", // should not add anything when there is nothing to add
      "animateWithDuration", // should only use label unless all other labels are identical
      "animateWithDurationDelay", // should only use labels when last label is unique
      "animateWithDurationDelayEasing"
    });
  }

  public void test_duplicateMethodName_shouldUseTypes_whenLabelsMatch() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("setValue", "_ value: String"),
      new MethodModel("setValue", "_ value: Int"),
      new MethodModel("set", "object: String", "forKey key: String"),
      new MethodModel("set", "object: Int", "forKey key: String"),
      new MethodModel("setNumber", "_ number: Float", "at index: Int"),
      new MethodModel("setNumber", "_ number: Int", "forKey key: String"),
      new MethodModel("setMultiple", "_ number: Int", "for key: Int"),
      new MethodModel("setMultiple", "_ number: Int", "for key: String")
    }, new String[]{
      // should use type when method name and parameters are overloaded
      "setValueString",
      "setValueInt",
      "setObjectStringForKey",
      "setObjectIntForKey",
      "setNumberAt",
      "setNumberForKey",
      "setMultipleIntForInt",
      "setMultipleIntForString",
    });
  }

  public void test_duplicateMethodName_shouldUseTypes_whenLabelsMatch_andNextParamsMatch() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("setValue", "_ value: String"),
      new MethodModel("setValue", "_ value: Int"),
      new MethodModel("setValue", "_ value: String", "forKey key: String"),
      new MethodModel("setValue", "_ value: Int", "forKey key: String")
    }, new String[]{
      // should use type when labels match and there is another identical method except its type
      "setValueString",
      "setValueInt",
      "setValueStringForKey",
      "setValueIntForKey",
    });
  }

  public void test_shouldIgnoreDefaultArguments() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("method", "param: String = \"\""),
      new MethodModel("method", "param: Int = 345")
    }, new String[]{
      "methodParamString",
      "methodParamInt",
    });
  }

  public void test_shouldProcessOneLetterMethodNames() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("a", ""),
      new MethodModel("a", "b")
    }, new String[]{
      "a",
      "aB",
    });
  }

  public void test_shouldAllowDuplicateMethods() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("method"),
      new MethodModel("method")
    }, new String[]{
      "method",
      "method"
    });
  }

  public void test_shouldProcessStrangeWhitespace() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("method", "    param1     :     String   ", " param3   label    : Int  "),
      new MethodModel("method", "param1:String", "param2:Int")
    }, new String[]{
      "methodParam1Param3",
      "methodParam1Param2"
    });
  }

  public void test_shouldProcessIncompleteParameters() throws Exception {
    assertEquals(new MethodModel[]{
      new MethodModel("method", "_"),
      new MethodModel("method", "param1"),
      new MethodModel("method", ":String"),
      new MethodModel("method", ":Int")
    }, new String[]{
      "method",
      "methodParam1",
      "methodString",
      "methodInt"
    });
  }

  private UniqueMethodNameGenerator createGenerator(MethodModel... models) {
    generator = new UniqueMethodNameGenerator(models);
    generator.generateMethodNames();
    return generator;
  }

  private void assertEquals(MethodModel[] models, String[] expected) {
    createGenerator(models);
    for (int i = 0; i < models.length; i++) {
      assertEquals(expected[i], generator.getMethodName(models[i].getID()));
    }
  }
}

File: UseCases/tests/codes/seanhenry/mockgenerator/util/UniqueMethodNameGeneratorTests.kt
package codes.seanhenry.mockgenerator.util

import junit.framework.TestCase
import kotlin.test.assertEquals
import kotlin.test.assertNull

class UniqueMethodNameGeneratorTests : TestCase() {

  private lateinit var generator: UniqueMethodNameGenerator

  fun test_uniqueMethodName_shouldReturnMethodName() {
    assertEquals(arrayOf(MethodModel("methodName")), arrayOf("methodName"))
  }

  fun test_shouldReturnNull_whenIDDoesNotExist() {
    generator = createGenerator()
    assertNull(generator.getMethodName("1"))
  }

  fun test_uniqueMethodName_shouldReturnMethodName_whenNoOverloadedMethods() {
    assertEquals(arrayOf(
        MethodModel("methodName", "param", "param2")
    ), arrayOf(
        "methodName" // should not add labels when method name is unique
    ))
  }

  fun test_duplicateMethodName_shouldAppendFirstParameterLabel_toMethodName() {
    assertEquals(arrayOf(
        MethodModel("methodName"),
        MethodModel("anotherMethod"),
        MethodModel("methodName", "param: Type")
    ), arrayOf(
        "methodName",
        "anotherMethod",
        // should add label when method name is overloaded
        // should not use type when label is unique
        "methodNameParam"
    ))
  }

  fun test_duplicateMethodName_shouldAppendFirstParameterLabel_toMethodName_whenGivenMethod_hasMultipleParamLabels() {
    assertEquals(arrayOf(
        MethodModel("animate"),
        MethodModel("animate", "withDuration duration: Type"),
        MethodModel("animate", "withDuration duration: Type", "delay: Type"),
        MethodModel("animate", "withDuration duration: Type", "delay: Type", "easing: Ease")
    ), arrayOf(
        "animate", // should not add anything when there is nothing to add
        "animateWithDuration", // should only use label unless all other labels are identical
        "animateWithDurationDelay", // should only use labels when last label is unique
        "animateWithDurationDelayEasing"
    ))
  }

  fun test_duplicateMethodName_shouldUseTypes_whenLabelsMatch() {
    assertEquals(arrayOf(
        MethodModel("setValue", "_ value: String"),
        MethodModel("setValue", "_ value: Int"),
        MethodModel("set", "object: String", "forKey key: String"),
        MethodModel("set", "object: Int", "forKey key: String"),
        MethodModel("setNumber", "_ number: Float", "at index: Int"),
        MethodModel("setNumber", "_ number: Int", "forKey key: String"),
        MethodModel("setMultiple", "_ number: Int", "for key: Int"),
        MethodModel("setMultiple", "_ number: Int", "for key: String")
    ), arrayOf(
        // should use type when method name and parameters are overloaded
        "setValueString",
        "setValueInt",
        "setObjectStringForKey",
        "setObjectIntForKey",
        "setNumberAt",
        "setNumberForKey",
        "setMultipleIntForInt",
        "setMultipleIntForString"
    ))
  }

  fun test_duplicateMethodName_shouldUseTypes_whenLabelsMatch_andNextParamsMatch() {
    assertEquals(arrayOf(
        MethodModel("setValue", "_ value: String"),
        MethodModel("setValue", "_ value: Int"),
        MethodModel("setValue", "_ value: String", "forKey key: String"),
        MethodModel("setValue", "_ value: Int", "forKey key: String")
    ), arrayOf(
        // should use type when labels match and there is another identical method except its type
        "setValueString",
        "setValueInt",
        "setValueStringForKey",
        "setValueIntForKey"
    ))
  }

  fun test_shouldIgnoreDefaultArguments() {
    assertEquals(arrayOf(
        MethodModel("method", "param: String = \"\""),
        MethodModel("method", "param: Int = 345")
    ), arrayOf(
        "methodParamString",
        "methodParamInt"
    ))
  }

  fun test_shouldProcessOneLetterMethodNames() {
    assertEquals(arrayOf(
        MethodModel("a", ""),
        MethodModel("a", "b: Type")
    ), arrayOf(
        "a", "aB"
    ))
  }

  fun test_shouldAllowDuplicateMethods() {
    assertEquals(arrayOf(
        MethodModel("method"),
        MethodModel("method")
    ), arrayOf(
        "method",
        "method"
    ))
  }

  fun test_shouldProcessStrangeWhitespace() {
    assertEquals(arrayOf(
        MethodModel("method", "    param1     :     String   ", " param3   label    : Int  "),
        MethodModel("method", "param1:String", "param2:Int")
    ), arrayOf(
        "methodParam1Param3",
        "methodParam1Param2"
    ))
  }

  fun test_shouldIgnoreIncompleteParameters() {
    assertEquals(arrayOf(
        MethodModel("method", "_"),
        MethodModel("method", "param1"),
        MethodModel("method", ":String"),
        MethodModel("method", ":Int")
    ), arrayOf(
        "method",
        "method",
        "method",
        "method"
    ))
  }

  fun test_shouldIgnoreSpecialCharactersInTuplesAndClosures() {
    assertEquals(arrayOf(
        MethodModel("method", "_ tuple: (String, Int)"),
        MethodModel("method", "_ tuple: (UInt, Float)"),
        MethodModel("anotherMethod", "_ closure: (Int) -> ()"),
        MethodModel("anotherMethod", "_ closure: () -> String ")
    ), arrayOf(
        "methodStringInt",
        "methodUIntFloat",
        "anotherMethodInt",
        "anotherMethodString"
    ))
  }

  private fun createGenerator(vararg models: MethodModel): UniqueMethodNameGenerator {
    generator = UniqueMethodNameGenerator(*models)
    generator.generateMethodNames()
    return generator
  }

  private fun assertEquals(models: Array<MethodModel>, expected: Array<String>) {
    createGenerator(*models)
    for (i in models.indices) {
      assertEquals(expected[i], generator.getMethodName(models[i].id))
    }
  }
}

